### DISCUSSION TITLE
About pairing with point/mobile data vs gridded

### DISCUSSION THREAD

**Discussion by blychs** [NOAA-CSL/MELODIES-MONET#325] [General]
Hi all,
While working on a PR for arbitrary region selection and testing my code, I came across the following question, which I think merits discussion.
Currently, when we pair with point data, we save (sensibly, IMHO) only those points that are paired. This, however, has the following disadvantage: an.read_analysis() is not sufficient for all our plots. I.e., plots like the spatial_overlay() require reading also the model.
This is not necessarily a problem, it somehow makes sense that "paired_data" includes only those data that are effectively paired.
However, it leads to a difference between pairing with gridded data (i.e., satellite or, maybe in the future, pairing two models), where the paired.obj are self contained and we can do all plot just reading that in, and point/mobile data, where that is not the case, especially for maps. Right now, the fact that the model needs to be read again is not obvious in the docs nor in the examples.

I can think of 2 potential solutions:
1) Just updating the docs to make that difference clear and specify that, for some of the plots (spatial overlay being the only example right now) the user still needs to run an.read_model(). I would also add a clear error message to make_spatialoverlay if the user did not read the model after running an.read_analysis(), and probably deal with it by skipping it. It does not make sense that they would need to re-run everything if just those plots failed. **I'm leaning towards this, it just seems like a better, more straightforward solution.**
2) Changing the paired data to include all of it. This has quite some disadvantages, since it would mean having data full of NaN for the observations. It is also not clear how the resample would work for mobile/aircraft. It has the main advantage that the data saved with an.save_analysis() would be self contained.
@rschwant , @zmoon , @lkemmons what do you think?

**Comment by zmoon** [comment]
> only those points that are paired.

Technically I believe what we currently save[^a] includes all of the _obs_ time-space points. This can lead to data being saved that aren't actually paired, e.g. due to differences in time ranges or spatial coverage.

I think ideally the saved paired data should only include the useful paired points (time-space points that are non-null for both in the pair). But then the result of reading this is not exactly the same as the pairing result MM gives, currently at least.

I think making the situation more clear in the docs sounds good and improving error messages.

In principle, for `melodies-monet run` (CLI) it should be possible to detect if reading the model(s) is necessary or not when using saved paired data. But it could be a little tricky with multiple models/pairs (maybe you only want spatial overlay plot for one model, so you only have to read that one really).

[^a]: at least in my experience with surface obs vs model

**Reply by blychs** [reply]
I think that you are right, we kind of assume that the obs are contained in the model, which is not necessarily true. This does not happen in all of the satellite, where we often discard anything outside of the model domain, but I believe it happens elsewhere.
I do think that improved error messages and docs is the way to go.

**Reply by zmoon** [reply]
> the useful paired points (time-space points that are non-null for both in the pair)

Also, for an analysis situation with multiple pairs (e.g. multiple versions of a model compared to obs), I've found it useful to have a single combined paired data file with the time-space points that are non-null for _all_ data sources. It would be nice if MM could optionally produce that.
